cpmpy/solvers/scip.py

```diff
@@ -0,0 +1,419 @@
+
+    The `DirectConstraint`, when used, calls a function on the `scip_model` object.
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import pyscipopt as scip
+            return True
+        except:
+            return False
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: a CPMpy Model()
+        """
+        if not self.supported():
+            raise Exception(
+                "CPM_scip: Install SCIPOptSuite and the python package 'pyscipopt'!")
+        import pyscipopt as scip
+
+        self.scip_model = scip.Model("From CPMpy")
+        self.scip_model.setParam("display/verblevel", 0) # remove solver logs from output
+        # initialise everything else and post the constraints/objective
+        # it is sufficient to implement __add__() and minimize/maximize() below
+        super().__init__(name="scip", cpm_model=cpm_model)
+
+
+    def solve(self, time_limit=None, solution_callback=None, **kwargs):
+        """
+            Call the SCIP solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Arguments that correspond to solver parameters:
+            Examples of scip supported arguments include:
+                - TODO
+
+            For a full list of scip parameters, please visit 
+            https://www.scipopt.org/doc/html/PARAMETERS.php
+        """
+        #import pyscipopt as scip
+
+        if time_limit is not None:
+            # in seconds
+            self.scip_model.setParam("limits/time", time_limit)
+
+        if solution_callback is not None:
+            raise Exception("SCIP: solution callback not (yet?) implemented")
+
+        # call the solver, with parameters
+        #for param, val in kwargs.items():
+        #    self.scip_model.setParam(param, val)
+        self.scip_model.setParams(kwargs)
+
+        _ = self.scip_model.optimize()
+        scip_objective = self.scip_model.getObjective()
+
+        scip_status = self.scip_model.getStatus()
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = self.scip_model.getSolvingTime()
+
+        # translate exit status
+        if scip_status == "optimal":
+            if self.has_objective():
+                self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+            else:
+                self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif scip_status == "infeasible":
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif scip_status.endswith("limit"):  # timelimit, nodelimit, ...
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
```

Comment:
This should only be feasible if at least one solution is found, is there a way to check this? Otherwise should be set to unknown to avoid trying to print a solution that doesn't exist later on

Reply:
To check if a valid solution exists in storage you can call:
```if self.scip_model.getNSols() > 0```